src/main/java/com/michelin/ns4kafka/controllers/SchemaController.java

```diff
@@ -100,52 +101,63 @@ public Mono<HttpResponse<Schema>> apply(String namespace, @Valid @Body Schema sc
-                        schema.getMetadata().setNamespace(ns.getMetadata().getName());
-                        latestSubjectOptional.ifPresent(
-                            value -> schema.getSpec().setCompatibility(value.getSpec().getCompatibility()));
-
-                        if (dryrun) {
-                            // Cannot compute the "unchanged" apply status before getting the ID at registration
-                            return Mono.just(formatHttpResponse(schema,
-                                latestSubjectOptional.isPresent() ? ApplyStatus.changed : ApplyStatus.created));
+                        if (newSchema.canonicalString().equals(actualSchema.canonicalString())) {
+                            return Mono.just(formatHttpResponse(schema, ApplyStatus.unchanged));
                         }
 
-                        return schemaService
-                            .register(ns, schema)
-                            .map(id -> {
-                                ApplyStatus status;
-
-                                if (latestSubjectOptional.isEmpty()) {
-                                    status = ApplyStatus.created;
-                                    sendEventLog(schema.getKind(), schema.getMetadata(), status, null,
-                                        schema.getSpec());
-                                } else if (id > latestSubjectOptional.get().getSpec().getId()) {
-                                    status = ApplyStatus.changed;
-                                    sendEventLog(schema.getKind(), schema.getMetadata(), status,
-                                        latestSubjectOptional.get().getSpec(),
-                                        schema.getSpec());
-                                } else {
-                                    status = ApplyStatus.unchanged;
+                        schema.getSpec().setCompatibility(latestSubjectOptional.get().getSpec().getCompatibility());
+                    }
+
+                    return schemaService
+                            .validateSchemaCompatibility(ns.getMetadata().getCluster(), schema)
+                            .flatMap(validationErrors -> {
+                                if (!validationErrors.isEmpty()) {
+                                    return Mono.error(new ResourceValidationException(
+                                            validationErrors,
+                                            schema.getKind(),
+                                            schema.getMetadata().getName()));
+                                }
+
+                                schema.getMetadata().setCreationTimestamp(Date.from(Instant.now()));
+                                schema.getMetadata().setCluster(ns.getMetadata().getCluster());
+                                schema.getMetadata().setNamespace(ns.getMetadata().getName());
+
+                                if (dryrun) {
```

Comment:
I think this:

```java
if (dryrun) {
  return Mono.just(formatHttpResponse(schema,
          latestSubjectOptional.isPresent()
                  ? ApplyStatus.changed : ApplyStatus.created));
}
  
return schemaService
      .register(ns, schema)
      .map(id -> {
          ApplyStatus status;
  
          if (latestSubjectOptional.isEmpty()) {
              status = ApplyStatus.created;
              sendEventLog(schema.getKind(), schema.getMetadata(), status, null,
                  schema.getSpec());
          } else if (id > latestSubjectOptional.get().getSpec().getId()) {
              status = ApplyStatus.changed;
              sendEventLog(schema.getKind(), schema.getMetadata(), status,
                  latestSubjectOptional.get().getSpec(),
                  schema.getSpec());
          } else {
              status = ApplyStatus.unchanged;
          }
  
          return formatHttpResponse(schema, status);
      });
});
```

could be now simplified to this:

```java
ApplyStatus status = latestSubjectOptional.isPresent() ? ApplyStatus.changed : ApplyStatus.created;
if (dryrun) {
    return Mono.just(formatHttpResponse(schema, status);
}

return schemaService
    .register(ns, schema)
    .map(id -> formatHttpResponse(schema, status));
```

Since we now use the canonical string to compute the status, we do not need the schema registry response for the final HTTP response, what do you think ? I'm gonna try this out